File: 2026/08-August/19-Wednesday/solution.c
#include <stdlib.h>
#include <stdbool.h>

typedef struct {
    int row;
    int mask;
} Row;

int maxNumberOfFamilies(int n, int** reservedSeats, int reservedSeatsSize,
                        int* reservedSeatsColSize) {

    Row* rows = (Row*)calloc(reservedSeatsSize, sizeof(Row));
    int rowCount = 0;

    for (int i = 0; i < reservedSeatsSize; i++) {
        int r = reservedSeats[i][0];
        int c = reservedSeats[i][1];

        int index = -1;

        for (int j = 0; j < rowCount; j++) {
            if (rows[j].row == r) {
                index = j;
                break;
            }
        }

        if (index == -1) {
            index = rowCount++;
            rows[index].row = r;
            rows[index].mask = 0;
        }

        rows[index].mask |= (1 << c);
    }

    int ans = (n - rowCount) * 2;

    for (int i = 0; i < rowCount; i++) {
        int mask = rows[i].mask;

        bool left =
            !(mask & (1 << 2)) &&
            !(mask & (1 << 3)) &&
            !(mask & (1 << 4)) &&
            !(mask & (1 << 5));

        bool middle =
            !(mask & (1 << 4)) &&
            !(mask & (1 << 5)) &&
            !(mask & (1 << 6)) &&
            !(mask & (1 << 7));

        bool right =
            !(mask & (1 << 6)) &&
            !(mask & (1 << 7)) &&
            !(mask & (1 << 8)) &&
            !(mask & (1 << 9));

        if (left && right)
            ans += 2;
        else if (left || middle || right)
            ans += 1;
    }

    free(rows);
    return ans;
}

File: 2026/08-August/19-Wednesday/solution.py
class Solution:
    def maxNumberOfFamilies(self, n: int, reservedSeats: list[list[int]]) -> int:
        reserved = {}

        for r, c in reservedSeats:
            reserved.setdefault(r, set()).add(c)

        # Four possible groups of 4 seats:
        # Left:   2,3,4,5
        # Middle: 4,5,6,7
        # Right:  6,7,8,9

        ans = (n - len(reserved)) * 2

        for seats in reserved.values():
            left = all(c not in seats for c in [2, 3, 4, 5])
            middle = all(c not in seats for c in [4, 5, 6, 7])
            right = all(c not in seats for c in [6, 7, 8, 9])

            if left and right:
                ans += 2
            elif left or middle or right:
                ans += 1

        return ans
